src/main/java/org/jabref/gui/fieldeditors/LinkedFileViewModel.java

```diff
@@ -345,4 +352,113 @@ public void writeXMPMetadata() {
+        String suggestedName = getSuggestedFileName(suffix);
+        Path directory;
+        if (fileDirectories.isEmpty()) {
+            directory = null;
+        } else {
+            directory = fileDirectories.get(0);
+        }
+        final Path suggestDir = directory == null ? Paths.get(System.getProperty("user.home")) : directory;
+        return suggestDir.resolve(suggestedName);
+    }
+
+    private Optional<ExternalFileType> inferFileType(URLDownload urlDownload) {
+        Optional<ExternalFileType> suggestedType = inferFileTypeFromMimeType(urlDownload);
+
+        // If we did not find a file type from the MIME type, try based on extension:
+        if (!suggestedType.isPresent()) {
+            suggestedType = inferFileTypeFromURL(urlDownload.getSource().toExternalForm());
+        }
+        return suggestedType;
+    }
+
+    private Optional<ExternalFileType> inferFileTypeFromMimeType(URLDownload urlDownload) {
+        try {
+            // TODO: what if this takes long time?
+            String mimeType = urlDownload.getMimeType(); // Read MIME type
+            if (mimeType != null) {
+                LOGGER.debug("MIME Type suggested: " + mimeType);
+                return ExternalFileTypes.getInstance().getExternalFileTypeByMimeType(mimeType);
+            } else {
+                return Optional.empty();
+            }
+        } catch (IOException ex) {
+            LOGGER.debug("Error while inferring MIME type for URL " + urlDownload.getSource(), ex);
+            return Optional.empty();
+        }
+    }
+
+    private Optional<ExternalFileType> inferFileTypeFromURL(String url) {
+        String extension = DownloadExternalFile.getSuffix(url);
+        if (extension != null) {
+            return ExternalFileTypes.getInstance().getExternalFileTypeByExt(extension);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    private String getSuggestedFileName(String suffix) {
+        String plannedName = FileUtil.createFileNameFromPattern(databaseContext.getDatabase(), entry,
+                Globals.prefs.get(JabRefPreferences.IMPORT_FILENAMEPATTERN));
+
+        if (!suffix.isEmpty()) {
+            plannedName += "." + suffix;
+        }
+
+        /*
+        * [ 1548875 ] download pdf produces unsupported filename
+        *
+        * http://sourceforge.net/tracker/index.php?func=detail&aid=1548875&group_id=92314&atid=600306
+        * FIXME: rework this! just allow alphanumeric stuff or so?
+        * https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx#naming_conventions
+        * http://superuser.com/questions/358855/what-characters-are-safe-in-cross-platform-file-names-for-linux-windows-and-os
+        * https://support.apple.com/en-us/HT202808
+        */
+        if (OS.WINDOWS) {
+            plannedName = plannedName.replaceAll("\\?|\\*|\\<|\\>|\\||\\\"|\\:|\\.$|\\[|\\]", "");
```

Comment:
extract to a matcher?

Reply:
It turns out that this code is not needed since the `FileUtil` class already handles cleanup of file names